cmake/create_test.cmake

```diff
@@ -101,63 +101,31 @@ function(ginkgo_create_cuda_test test_name)
     add_test(NAME ${REL_BINARY_DIR}/${test_name} COMMAND ${TEST_TARGET_NAME})
 endfunction(ginkgo_create_cuda_test)
 
-function(ginkgo_create_hip_test_special_linkage test_name)
-    # use gcc to compile but use hip to link
-    file(RELATIVE_PATH REL_BINARY_DIR
-        ${PROJECT_BINARY_DIR} ${CMAKE_CURRENT_BINARY_DIR})
-    string(REPLACE "/" "_" TEST_TARGET_NAME "${REL_BINARY_DIR}/${test_name}")
-    add_executable(${TEST_TARGET_NAME} ${test_name}.cpp)
-    target_compile_features("${TEST_TARGET_NAME}" PUBLIC cxx_std_14)
-    # Fix the missing metadata when building static library.
-    if(GINKGO_HIP_PLATFORM MATCHES "hcc" AND NOT BUILD_SHARED_LIBS)
-        set_target_properties(${TEST_TARGET_NAME} PROPERTIES LINKER_LANGUAGE HIP)
-    endif()
-    target_include_directories("${TEST_TARGET_NAME}"
-        PRIVATE
-        "$<BUILD_INTERFACE:${Ginkgo_BINARY_DIR}>"
-        )
-    set_target_properties(${TEST_TARGET_NAME} PROPERTIES
-        OUTPUT_NAME ${test_name})
-    if (GINKGO_CHECK_CIRCULAR_DEPS)
-        target_link_libraries(${TEST_TARGET_NAME} PRIVATE "${GINKGO_CIRCULAR_DEPS_FLAGS}")
-    endif()
-    target_link_libraries(${TEST_TARGET_NAME} PRIVATE ginkgo GTest::Main GTest::GTest ${ARGN})
-    add_test(NAME ${REL_BINARY_DIR}/${test_name} COMMAND ${TEST_TARGET_NAME})
-endfunction(ginkgo_create_hip_test_special_linkage)
 
 function(ginkgo_create_hip_test test_name)
     file(RELATIVE_PATH REL_BINARY_DIR
         ${PROJECT_BINARY_DIR} ${CMAKE_CURRENT_BINARY_DIR})
     string(REPLACE "/" "_" TEST_TARGET_NAME "${REL_BINARY_DIR}/${test_name}")
 
     set_source_files_properties(${test_name}.hip.cpp PROPERTIES HIP_SOURCE_PROPERTY_FORMAT TRUE)
+    hip_add_executable(${TEST_TARGET_NAME} ${test_name}.hip.cpp
+        # With how HIP works, to prevent redifinition we need to comment these.
+        # They might need to be used later though.
+        #HIPCC_OPTIONS ${GINKGO_HIPCC_OPTIONS}
+        #NVCC_OPTIONS  ${GINKGO_HIP_NVCC_OPTIONS}
+        #CLANG_OPTIONS ${GINKGO_HIP_CLANG_OPTIONS}
```

Comment:
what do you mean redefinition here?
If it does not contain these option, can it get the arch flag and optimization?

Reply:
I have trouble explaining how it works but indeed with the current setting everything is passed correctly. I can only explain it as a bug in FindHIP.cmake, that bug would explain why we have so many duplications of all the flags as well.

For example:
```
/opt/rocm-4.0.0/hip/bin/hipcc -M /home/terry/ginkgo/hip/test/base/hip_executor_topology.hip.cpp -o /home/terry/ginkgo/build/hip/test/base/CMakeFiles/hip_test_base_hip_executor_topology.dir//hip_test_base_hip_executor_topology_generated_hip_executor_topology.hip.cpp.o.depend.pre -std=c++14 -fPIC --amdgpu-target=gfx906 --amdgpu-target=gfx908 -fPIC -O2 -DNDEBUG -I/home/terry/ginkgo/build -I/opt/rocm-4.0.0/include -I/opt/rocm-4.0.0/hipblas/include -I/opt/rocm-4.0.0/hiprand/include -I/opt/rocm-4.0.0/hipsparse/include -I/home/terry/ginkgo/build/include -I/home/terry/ginkgo/include -I/home/terry/ginkgo -I/usr/local/include -I/home/terry/ginkgo/build/include -I/home/terry/ginkgo/include -I/home/terry/ginkgo -I/usr/local/include -I/usr/local/include -I/home/terry/ginkgo/build/include -I/home/terry/ginkgo/include -I/home/terry/ginkgo -I/usr/local/include -I/home/terry/ginkgo/build/include -I/home/terry/ginkgo/include -I/home/terry/ginkgo -I/usr/local/include -I/home/terry/ginkgo/build/include -I/home/terry/ginkgo/include -I/home/terry/ginkgo -I/usr/local/include -I/opt/rocm-4.0.0/hip/include -I/home/terry/ginkgo/build/include -I/home/terry/ginkgo/include -I/home/terry/ginkgo -I/usr/local/include -I/home/terry/ginkgo/build/include -I/home/terry/ginkgo/include -I/home/terry/ginkgo -I/usr/local/include -I/home/terry/ginkgo/build/third_party/gtest/src/googletest/include -I/home/terry/ginkgo/build/third_party/gtest/src/googletest/include
```

Reply:
Note that If I pass the flags again (at least HIPCC_OPTIONS), there is two `std=c++14` and somehow that makes the command invalid. I can maybe reactivate the other flags at least for safety?

Reply:
I didn't try it to make 100% sure but I think the problem is due to these:
https://github.com/ROCm-Developer-Tools/HIP/blob/main/cmake/FindHIP.cmake#L463-L466
https://github.com/ROCm-Developer-Tools/HIP/blob/main/cmake/FindHIP.cmake#L381-L397

By looking only at these two, since no reset seem to be done, they *always* append the flags that they find to the *global* variables `HIP_HIPCC_FLAGS` and whatnot. The macro doing this is called whenever we call a `hip_add_library` or `hip_add_executable`. Thus, when we do `hip_add_executable` for the test, we already had the definitions from `libginkgo_hip.so` which get reused and thus we do not need to pass the options again (and it creates an error in the case of `-std=c++14`).
If they used a function that would make things much simpler and they could use a scoped variable to manage the flags that they find. Or they could just add them to the property of the object they are defining. I do not understand what they are doing.

In any case, I will try to put them back and see if it works, although that will duplicate everything at least twice.

Reply:
For completeness, here is the failing build, I think it's only `nvcc` which creates issues so I will make a special case out of it:
https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/1060533273
```
nvcc fatal   : redefinition of argument 'std'
```